dlhub_sdk/utils/validation.py

```diff
@@ -0,0 +1,239 @@
+
+    Args:
+        tup (tuple): The tuple whose items need to have their type validated
+        db_entries (list): The list of dicts that store the metadata for each item in tup
+        path (list): List that stores the path through the data to the current point
+    Returns:
+        None
+    Raises:
+        ValueError: If the number of entries and tuple elements do not match
+        TypeError: If any of the items' types do not match
+    """
+    if len(tup) != len(db_entries):
+        raise _generate_err(ValueError, path, msg=f"dl.run expected tuple of length {len(db_entries)}, recieved tuple with length {len(tup)}"
+                                                  + "{loc}")
+
+    path.append(["tuple", None])
+    for i, (given, expected) in enumerate(zip(tup, db_entries)):
+        path[-1][1] = i
+        validate(given, expected, path)
+    path.pop()
+
+
+def _validate_ndarray(arr: ndarray, shape: list, db_entry: dict, path: list) -> None:
+    """Compare the shape of arr with shape and recursively validate each of the elements in arr against db_entry["item_type"]
+
+    Args:
+        arr (ndarray): The ndarray whose shape and items need to be validated
+        shape (list): The shape that arr is expected to have
+        db_entry (dict): The entry where type data may or may not be retrieved
+        path (list): List that stores the path through the data to the current point
+    Returns:
+        None
+    Raises:
+        ValueError: If arr.shape does not match shape
+        TypeError: If any of the items' types do not match
+    """
+    shape = tuple(int(x) if x != "None" else x for x in shape)
+    shape_err = _generate_err(ValueError, path, msg=f"dl.run given improper input: expected ndarray.shape = {shape}, received shape: {arr.shape}"
+                                                    + "{loc}")
+
+    if len(arr.shape) != len(shape):
+        raise shape_err
+    if not all(map(lambda t: t[0] == t[1] or t[1] == "None", zip(arr.shape, shape))):
+        raise shape_err
+
+    entry = db_entry.get("item_type")
+    if entry:
+        path.append(["ndarray", None])
+        for i, item in ndenumerate(arr):
+            path[-1][1] = i
+            validate(item.item(), entry, path)
+        path.pop()
+
+
+def _validate_dict(dct: dict, props: dict, path: list) -> None:
+    """Recursively validate each of the pairs in dct against props
+
+    Args:
+        dct (dict): The dict whose keys and values need to be validated
+        props (dict): The dict that describes the expected keys and values
+        path (list): List that stores the path through the data to the current point
+    Returns:
+        None
+    Raises:
+        ValueError: If dct is missing a key from props or a value in dct causes an error
+        TypeError: If any of the values' types do not match
+    """
+    if props:
+        path.append(["dict", None])
```

Comment:
As above, add some line comments

Reply:
e.g., the first loop is for detecting unauthorized keys and the second is looking for required keys.